File: Data Structures & Algorithms/target-sum/submission-3.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        
        memo = {}

        def dfs(i, acc):
            if i == len(nums):
                if acc == target:
                    memo[(i, acc)] = 1
                    return 1
                else:
                    memo[(i, acc)] = 0
                    return 0
            if (i, acc) in memo:
                return memo[(i, acc)]
            
            sum_plus = dfs(i + 1, acc + nums[i])
            sum_minus = dfs(i + 1, acc - nums[i])

            memo[(i, acc)] = sum_plus + sum_minus 

            return sum_plus + sum_minus 
        
        return dfs(0, 0)

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-3.java
class Solution {
    public int search(int[] nums, int target) {
        
        int l = 0;
        int r = nums.length - 1;

        while (l <= r) {
            int m = (l + r) / 2;
            if (nums[m] == target) {
                return m;
            }

            if (nums[m] >= nums[0]){
                // sorted left half 
                if (target < nums[m]) {
                    if (target < nums[0]){
                        l = m + 1;
                    } else {
                        r = m - 1;
                    }
                } else {
                    l = m + 1; 
                }
            } else {
                if (target > nums[m]){
                    if (target > nums[nums.length - 1]) {
                        r = m - 1;
                    } else {
                        l = m + 1;
                    }
                } else {
                    r = m - 1;
                }

            }
        }

        return -1;
    }
}
